### ISSUE TITLE
All() should return a tuple instead of positional arguments

### ISSUE THREAD

**Issue by ppalmes** [JuliaData/DataFrames.jl#2439]
- All() operates on collection of columns and typically, you need to operate in many columns so it is easier to process vector or tuple of rows instead of using positional arguments.
- All() -> ByRow(sum or mean) won't work including other statistical metrics because All returns positional arguments and most functions process either vectors or tuples.
- All()=>ByRow((x...) -> sum(x)) or AsTable(All()) => (x->sum(x)) is needed but it is messy because why do you have to transform All() into a table because from the user perspective, dataframe is a table already. It is not clear why you need AsTable with All(). This happens because All() returns a positional value instead of a collection like vector or tuple
- most functions return scalar, or tuple, or vector. it is confusing to find All() returning positional argument for typical user. it is convenient to deal with two or three columns but many use-cases require operating each row across several columns

**Comment by bkamins** [comment]
> `All() -> ByRow(sum or mean)` won't work

`All() => ByRow(+)` works, as for mean indeed you have to write `combine(df, AsTable(All()) => ByRow(mean))`

> `AsTable(All()) => (x->sum(x))` is needed

It is enough to write:
```
All() => +
```
or
```
AsTAble(All()) => sum
```
in this case

> but many use-cases require operating each row across several columns

As a side note. These operations will be inefficient using a `DataFrame`, if you have such a use case probably a different data structure should be used (DataFrames.jl is optimized for column access not for row access).

---

Having said that, we could have another wrapper similar to `AsTable`, e.g. `AsTuple` that would produce a `Tuple` instead.

**Comment by pdeffebach** [comment]
Thanks for the issue report! 

I acknowledge that `All()` is awkward in this context since it has an undefined number of inputs so it mirrors splatting more. However note that

1. Plenty of functions allow `spat`ting, i.e. `+`, `string`, `*`, so these things will just work with current behavior.
2. I disagree that "working with many many columns" is the dominant use-case. See some code I wrote for a personal project [here](https://gitlab.com/pdeffebach/senate-voting/-/blob/master/src/import_clean_funs.jl). Plenty of uses of `transform` with positional arguments. 
3.   > into a table because from the user perspective, dataframe is a table already. It is not clear why you need AsTable with All(). This happens because All() returns a positional value instead of a collection like vector or tuple

note that you aren't advocating making a DataFrame the default output of `All()`, so this problem wouldn't be fixed by your proposal. 

4. It is definitely confusing that `All()` doesn't return a "thing". I agree, but we are trying to emulate the anonymous function syntax here a little bit, i.e. `[:x, :y] => fun` is "kind of" like `(x, y) -> fun`. It's not 1:1 but thats the mental model I think users should have. 

5. Note that all the functionality you request is currently available in `AsTable`. So these are things you *can* currently do. I presume you would like `AsTable` to return a `DataFrame`. 

If you are really working with 100s of columns, then `AsTable` might be slow due to compilation. I think we should maybe have an `AsDataFrame` type for working with many many columns. But then you might as well just not do piping all together. That's always an option.

**Comment by ppalmes** [comment]
i think the documentation can also be improved on this. because when i first tried the examples, i really am confused why i selected columns and then do some summary statistics to them and then have the error that sum/mean/median/ etc cannot operate on those columns you selected. it is typical for example that you iterate on df based on certain groupings, do transformations column-wise or row-wise, and expect that these operations have similar view whether you are doing it columnwise or rowwise from the user perspective. i get a row, i expect that i can just get the mean elements of this slice. my understanding is if you get a slice of table, you get a vector-like data you can operate. for example, it will be normal to do All(r"x*") or r"x*" to gather all x's to get mean of them. with positional args, you have to do (x...)->f(x) for most of aggregation functions. i think the use-case of + or * are less than many of aggregation functions for vectors.

**Comment by ppalmes** [comment]
a good use case is that you x1,..., x100 columns each are sensor data located in certain regions. it is time-series so your rows are time for example. to get an average sensor value per hour or time unit, you need to get the mean per row of all those x's but you need to use regular expressions because another region y's have also similar data in the same dataframe.

**Comment by ppalmes** [comment]
I just think that AsTuple or AsVector should be hidden away from ordinary user. they should expect the regular expression to return something that behaves like a table or elements of table which is a vector. by having that function, the user has to know the difference between positional thing vs tuple vs vector. They can be added but I'm thinking of the default behavior when you slice a column or select a column by regular expression. for df[:, r"x"], i expect a dataframe? but if i do transform(df, r"x"......) the r"x" passed to anonymous function is not a tuple/vector which is the one that confuses me or maybe others.

**Comment by ppalmes** [comment]
ByRow(max) works but ByRow(sum) which looks similar because they aggregate things don't work. both says, get the rows, sum them. get the rows, give me the max value. this is what the user thinks.

**Comment by bkamins** [comment]
As @nalimilan - this was the default in the past. We have changed it because of compilation latency.

What we have now was selected as a default because it reduces compilation times in the most common cases (and there were user reports that using what we had as a legacy solution caused so many recompilations that it became unacceptable).

So the default behavior is something to be learned unfortunately.

However, as said - I am open to having different inputs to functions using special wrappers.

**Comment by ppalmes** [comment]
yeah, i think we have to hide away the complexity from the user. i think it's just a matter of what is expected by the user who is not familiar with the existence of special commands like AsTuple or AsVector. regular expression is really good in slicing the columns so ByRow(fun) confuses me because it does not deal in collections of elements.

**Comment by pdeffebach** [comment]
I agree that it would make sense to have `AsTable` be the default. But then we would have users complain about unnecessary indexing for functions with positional arguments. We would have to introduce a `AsPositional` term. 

I think what people find the default and intuitive is very subjective. Its a tough balancing act.

**Comment by ppalmes** [comment]
the errors can be intimidating by missing the AsTable() thing. i think positional thing is less common in terms of expectations. seldom i encounter function that returns not a scalar value, tuple, or vector.

**Comment by ppalmes** [comment]
collection of positional thing as returned value is the main thing that confused me the most. i did not expect that return value.

**Comment by pdeffebach** [comment]
As noted above, with data cleaning you absolutely want positional arguments a lot. Many of us come from a social science background, which probably lead to that being the default choice.

**Comment by ppalmes** [comment]
i speak in general about return values you expect in julia. most functions deal with vectors, tuples specially if you do regular expressions which means it can return a lot of values that's why you do regex

**Comment by ppalmes** [comment]
you don't need regex if you operate in 2 or 3 columns

**Comment by ppalmes** [comment]
my expectations is that All() or r"x*" means that you are really dealing with loads of columns that's why you need regex

**Comment by ppalmes** [comment]
i don't really have use-case. i was just playing with the latest documentation and just checking out if my expectations of the behaviour of those examples are correct. it just stumped me when i have error like ByRow(sum)

**Comment by bkamins** [comment]
> seldom i encounter function that returns not a scalar value, tuple, or vector.

Just to be clear `All()` or `r"x"` do not return anything. They are just selectors.

>  `All()` or `r"x*"` means that you are really dealing with loads of columns

Agreed, but the fact that they work is a side effect of the flexibility of the design, not the main use case for with the current API was designed, which is more like `[:x, :y] => cor`, or `[:x, :y] => string` (not many columns).

Again - let me stress - it is inefficient to process many columns row-wise in `DataFrame` - this is not a right container for such a case, as `DataFrame` is designed for column-wise operations. For many rows - better collect these rows in to row-oriented storage type.

**Comment by ppalmes** [comment]
to me it's more of the consistency of expectations. if you use regex or all(), expect any number of columns so the logical expectation is if you do ... => x->f(x), x is a collection not positional. i think adding an API which is confusing is less helpful. if you just want to deal with :x,:y, don't introduce All() or r"x*" which become confusing in terms of consistency but this is just my opinion. users have no idea that r"x*" is not really a slice but just description because when you do df[:,r"x*"] it gives you the subdataframe.

**Comment by nalimilan** [comment]
Nobody likes confusing APIs. The problem is that your expectations are not everybody's expectations. "Consistency" means "do the same thing everywhere", and that's precisely what we do -- it doesn't mean "do what I expect" as the computer doesn't know that.

**Comment by ppalmes** [comment]
anyway, this is my feedback after trying the API. i hope my expectation is not common.

**Comment by bkamins** [comment]
And this is appreciated. Actually what you describe is what we had in the past and people were complaining about:
1. usefulness of what is returned
2. compilation latency

**Comment by ppalmes** [comment]
you use regular expression because you expect to avoid listing loads of them. i expect a collection. i don't know if this expectation is wrong. maybe, the documentation should explain more what's going on when you use All() and r"x*" to avoid guessing.

**Comment by ppalmes** [comment]
why not make All() and r"x*"  return AsTable and create a different API for positional? All() and r"x*" shout collections and I'm not sure why you need positional for the r"x*" expression.

**Comment by pdeffebach** [comment]
> why not make All() and r"x*" return AsTable and create a different API for positional? All() and r"x*" shout collections and I'm not sure why you need positional for the r"x*" expression.
> 
> the logic is i will use regex because i don't want to list manually the names, then it returns positional which means i need to list the names in the transformation. the regex and positional requirements are not compatible.

This is not a bad idea, but it's an inconsistency that would also be confusing. We have `AsTable` syntax for a reason. It's better users be explicit than have context-dependent behavior. 

These are tough design trade-offs. I think the expliciteness is very valuable here, but can maybe be played a bit looser with DataFramesMeta.

**Comment by bkamins** [comment]
@ppalmes - thank you for digging into it as it uncovered https://github.com/JuliaData/DataFrames.jl/issues/2440 which we have to carefully consider.

**Comment by ppalmes** [comment]
i'm glad a different issue was resolved because i was arguing more on my feeling of awkwardness of API which somehow just happened to coincide with the way the compiler behave in an awkward manner too for the positional values returned ;)

**Comment by bkamins** [comment]
Thank you for the contribution. As discussed, most likely for faster operation on rows I will propose `AsVector` or `AsTuple` (or both) wrappers, depending on how they turn out to work in tests.

**Comment by bkamins** [comment]
As a general comment - we will add functionality to DataFrames.jl that makes it efficient for different use cases, but we are in the stage where breaking changes will likely be rejected.

On the other hand "utility packages" - and there are many of them - are encouraged to experiment with different syntaxes, that will use low-level API of DataFrames.jl (and that is why we will work to provide efficient low-level building blocks for different use cases).
